crates/extensions/tedge-p11-server/src/pkcs11/mod.rs

```diff
@@ -201,21 +235,128 @@ impl Cryptoki {
         let token_info = context.get_token_info(slot)?;
         debug!(?slot_info, ?token_info, "Selected slot");
 
-        let session = context.open_ro_session(slot)?;
+        let session = match session_type {
+            CryptokiSessionType::ReadOnly => context.open_ro_session(slot)?,
+            CryptokiSessionType::_ReadWrite => context.open_rw_session(slot)?,
+        };
+
         session.login(UserType::User, Some(&self.config.pin))?;
         let session_info = session.get_session_info()?;
         debug!(?session_info, "Opened a readonly session");
 
+        let session = CryptokiSession {
+            session,
+            token_info,
+            uri_attributes,
+        };
+
+        let template = [];
+        let objects = session.session.find_objects(&template);
+        match objects {
+            Err(err) => {
+                error!(?template, ?err, "failed to find objects");
+            }
+            Ok(objects) => {
+                let objects = objects
+                    .into_iter()
+                    .flat_map(|o| {
+                        let uri = session.export_object_uri(o).inspect_err(
+                            |err| error!(?err, object = ?o, "failed to read properties of object"),
+                        );
+                        uri.map(|u| (o, u)).ok()
+                    })
+                    .collect::<Vec<_>>();
+                trace!(?objects, "Objects found in the token");
+            }
+        }
+
         Ok(session)
     }
 
-    pub fn signing_key(&self, uri: Option<&str>) -> anyhow::Result<Pkcs11Signer> {
-        let uri_attributes = self.request_uri(uri)?;
-        let session = self.open_session(&uri_attributes)?;
+    fn request_uri<'a>(
+        &'a self,
+        request_uri: Option<&'a str>,
+    ) -> anyhow::Result<uri::Pkcs11Uri<'a>> {
+        let mut config_uri = self
+            .config
+            .uri
+            .as_deref()
+            .map(|u| uri::Pkcs11Uri::parse(u).context("Failed to parse config PKCS#11 URI"))
+            .transpose()?
+            .unwrap_or_default();
+
+        let request_uri = request_uri
+            .map(|uri| uri::Pkcs11Uri::parse(uri).context("Failed to parse PKCS #11 URI"))
+            .transpose()?
+            .unwrap_or_default();
+
+        config_uri.append_attributes(request_uri);
+        Ok(config_uri)
+    }
+}
+
+fn get_all_slots_info(cryptoki: &Pkcs11) -> Vec<SlotInfo> {
+    let slots = match cryptoki.get_all_slots() {
+        Ok(slots) => slots,
+        Err(err) => {
+            error!(?err, "failed to get slots");
+            return vec![];
+        }
+    };
+    let slotinfos = slots
```

Comment:
Clippy warning

```
warning: returning the result of a `let` binding from a block
   --> crates/extensions/tedge-p11-server/src/pkcs11/mod.rs:316:5
```

Reply:
fixed in 014adb5

they must've disabled this lint in later versions because I'm developing on the current stable and I didn't got it.